src/Workspaces/Core/MSBuild.BuildHost/BuildHost.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Locator;
+using Microsoft.Build.Logging;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.MSBuild.Build;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+
+namespace Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost;
+
+internal sealed class BuildHost : IBuildHost
+{
+    private readonly JsonRpc _jsonRpc;
+    private readonly string? _binaryLogPath;
+
+    private readonly object _gate = new object();
+    private ProjectFileLoaderRegistry? _projectFileLoaderRegistry;
+    private ProjectBuildManager? _buildManager;
+
+    public BuildHost(JsonRpc jsonRpc, string? binaryLogPath)
+    {
+        _jsonRpc = jsonRpc;
+        _binaryLogPath = binaryLogPath;
+    }
+
+    [MemberNotNull(nameof(_projectFileLoaderRegistry), nameof(_buildManager))]
+    private void EnsureMSBuildLoaded(string projectFilePath)
+    {
+        lock (_gate)
+        {
+            // If we've already created our MSBuild types, then there's nothing further to do.
+            if (_buildManager != null && _projectFileLoaderRegistry != null)
+                return;
+
+#if NETFRAMEWORK
+
+            // In this case, we're just going to pick the highest VS install on the machine, in case the projects are using some newer
+            // MSBuild features. Since we don't have something like a global.json we can't really know what the minimum version is.
+
+            // TODO: we should also check that the managed tools are actually installed
+            MSBuildLocator.RegisterInstance(MSBuildLocator.QueryVisualStudioInstances().OrderByDescending(vs => vs.Version).First());
+
+#else
+
+            // Locate the right SDK for this particular project; MSBuildLocator ensures in this case the first one is the preferred one.
+            // TODO: we should pick the appropriate instance back in the main process and just use the one chosen here.
+            var options = new VisualStudioInstanceQueryOptions { DiscoveryTypes = DiscoveryType.DotNetSdk, WorkingDirectory = Path.GetDirectoryName(projectFilePath) };
```

Comment:
we might want to add some trace logging here to figure out what its actually loading (in case things go wrong).

Though I'm not sure exactly how we want to log from here.  We could send messages back to the main process via jsonrpc, or we could have a file logger that writes to the vscode extension log dir.  Up to you

Reply:
I think for logging I might just write to stderr; I think some of the libraries that MSBuildLocator calls might write to there directly.

Reply:
Works for me - we should definitely collect that in the host process and write it somewhere vscode can reach it (either the ouput window or a separate log file)

Reply:
Logging added!